api/envoy/config/core/v3/base.proto

```diff
@@ -444,3 +444,19 @@ message ControlPlane {
   // the Envoy is connected to.
   string identifier = 1;
 }
+
+// Configuration of DNS Lookup option flags which control the behavior of the DNS resolver.
+message DnsLookupOptions {
+  // Always use TCP queries instead of UDP queries for DNS lookups.
+  // This may be overridden on a per-cluster basis in cds_config,
+  // when :ref:`dns_resolvers <envoy_api_field_config.cluster.v3.Cluster.dns_resolvers>` and
+  // :ref:`use_tcp_for_dns_lookups <envoy_api_field_config.cluster.v3.Cluster.use_tcp_for_dns_lookups>` are
+  // specified.
+  // Setting this value causes failure if the
+  // ``envoy.restart_features.use_apple_api_for_dns_lookups`` runtime value is true during
+  // server startup. Apple's API only uses UDP for DNS resolution.
```

Comment:
@htuch this new message replaces a deprecated field (`use_tcp_for_dns_lookups`). However, it also modifies the field type from `bool` to `google.protobuf.BoolValue`.
I'm assuming it is ok, as the other field is deprecated, just want your opinion on this.

Reply:
I think this is fine. What is the default value for this field? `BoolValue` is mostly useful when we anticipate wanting to change the default (in particular for security related fields).

Reply:
The default value is protobuf's bool default value (`false`).
Actually, the name of the field `use_tcp_for_dns_lookups` implies it is meant to be "enabled". I'm not certain what's the added value of `BoolValue` in this case.

Reply:
So here is my approach to keep supporting the deprecated field "bool use_tcp_for_dns_lookups". I would love to get your opinions on them.

Even though the field "**_bool use_tcp_for_dns_lookups_**" is marked deprecated, we expect previous implementations of xDSv3 control plane to still set this value. So the logic will check if the new field "_google.protobuf.BoolValue use_tcp_for_dns_lookups_" is being set via _dns_lookup_options_ or not. The code will rely on **_has_**__use_tcp_for_dns_lookups()_ to check if this value is being programmed by the server. If this _dns_lookup_options_**_.has_use_tcp_for_dns_lookups()_** is false (which means value is not set) then the code will check if the deprecated field **_bool use_tcp_for_dns_lookups_** has the value **_true_** and uses it. This makes sure of the backward compatibility of the changes. 

Correct me if I am wrong here, by just using bool field type we will not be able to know if the config (true or false) is applied or not.

One might ask can we just do **_OR_** of **_bool use_tcp_for_dns_lookups_** and **_dns_lookup_options.use_tcp_for_dns_lookups().value()_**, like -> `config.dns_lookup_options.use_tcp_for_dns_lookups().value() || config.use_tcp_for_dns_lookups`. But as per the logic which is being worked upon now, the new field
**_google.protobuf.BoolValue dns_lookup_options.use_tcp_for_dns_lookups_**  will get preference and override any value set by bare **_bool use_tcp_for_dns_lookups_** value.

Reply:
I think you can use the presence of the `dns_lookup_options` message alone to detect that we aren't dealing with legacy configuration..

Reply:
Thanks for the suggestions. Not using BoolValue.